### PR TITLE
Tammy/add fm pileup aggregation

### PR DESCRIPTION
**Pull request by TammushB**

### CODE REVIEW

**Review by ilyasoifer — COMMENTED** [review]
I think it is OK to write it this way, the only preference is that the filtering (applying xgboost) and aggregation of the pileup should be separated.

**Review by GatKrieger — CHANGES_REQUESTED** [review]

File: src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+    "max_softclip_length",
+    "X_FLAGS",
+    "ML_QUAL",
+]
+columns_for_min_aggregation = ["X_QUAL", "X_INDEX"]
+columns_for_max_aggregation = ["X_QUAL", "X_INDEX"]
+columns_for_fillna = ["is_cycle_skip"]
+columns_for_st_et_aggregation = ["st", "et"]
+
+added_agg_features = {
```
[the remark this answers is not in the document]
Comment (TammushB) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py:77]: converted to 2 dicts

File: src/featuremap/ugbio_featuremap/featuremap_xgb_training.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    FeatureMapFields.PREV_3.value,
+    FeatureMapFields.NEXT_1.value,
+    FeatureMapFields.NEXT_2.value,
+    FeatureMapFields.NEXT_3.value,
+    FeatureMapFields.IS_CYCLE_SKIP.value,
+]
+custom_info_fields.extend(list(added_agg_features))
+
+
+def split_data_every_2nd_variant(X, y):  # noqa: N802, N803
```
[the remark this answers is not in the document]
Comment (GatKrieger) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_training.py:44]: Or, I guess it's better to define minimal/maximal number of variants to train on?

File: src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        agg_colname = colname.upper() + "_min"
+        record_dict_for_xgb[agg_colname] = min(record_info_dict[colname])
+    for colname in columns_for_max_aggregation:
+        agg_colname = colname.upper() + "_max"
+        record_dict_for_xgb[agg_colname] = max(record_info_dict[colname])
+    record_dict_for_xgb["count_forward"] = record_info_dict["is_forward"].split("|").count("T")
+    record_dict_for_xgb["count_reverse"] = record_info_dict["is_forward"].split("|").count("F")
+    record_dict_for_xgb["count_duplicate"] = record_info_dict["is_duplicate"].split("|").count(True)
+    record_dict_for_xgb["count_non_duplicate"] = record_info_dict["is_duplicate"].split("|").count(False)
+    if "is_cycle_skip" in record_info_dict:
```
[the remark this answers is not in the document]
Comment (TammushB) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py:156]: sorry can't take this without the 'if' statement. 
is_cycle_skip is not always present in the record_info_dict. 
So I need the if to add it if not present.

File: src/featuremap/ugbio_featuremap/featuremap_xgb_training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TammushB) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_training.py:38]: added another split option

File: src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    "X_INDEX_max": ["max value of X_INDEX", "Integer"],
+    "X_QUAL_min": ["min value of X_QUAL", "Float"],
+    "X_INDEX_min": ["min value of X_INDEX", "Integer"],
+    "count_forward": ["number of forward reads", "Integer"],
+    "count_reverse": ["number of reverse reads", "Integer"],
+    "count_duplicate": ["number of duplicate reads", "Integer"],
+    "count_non_duplicate": ["number of non-duplicate reads", "Integer"],
+}
+
+ppm_added_agg_features = {}
```
Comment (GatKrieger) [inline src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py:103]: also here..

File: src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (GatKrieger) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_prediction.py:148]: but the .get method deals with these cases. This is also what I did in: /home/ubuntu/ugbio-utils/src/featuremap/ugbio_featuremap/featuremap_consensus_utils.py line: 312

File: src/featuremap/ugbio_featuremap/featuremap_xgb_training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (GatKrieger) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_training.py:38]: So I guess there's no need in this function, right? this is just a specific case: split_data(X, y, test_size=0.5, random_state=42)

**Review by GatKrieger — APPROVED** [review]
I approve but I also ask for some changes.
@ilyasoifer please review the training part. I've read it but I'm not an expert (yet!)

**Review by ilyasoifer — APPROVED** [review]
Looks good, just minor things

File: src/featuremap/ugbio_featuremap/featuremap_xgb_training.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        X[col] = X[col].astype("category")
+
+    if args_in.split_data_every_2nd_variant:
+        [X_train, X_test, y_train, y_test] = split_data_every_2nd_variant(X, y)  # noqa: N806
+    elif args_in.test_size < 1 and args_in.test_size > 0:
+        [X_train, X_test, y_train, y_test] = split_data(X, y, test_size=args_in.test_size)  # noqa: N806
+    xgb_clf_es = XGBoost_train(X_train, y_train)
+    y_pred_es = XGBoost_test(xgb_clf_es, X_test)
+
+    XGBoost_evaluate_model(xgb_clf_es, y_test, y_pred_es)
```
[the remark this answers is not in the document]
Comment (TammushB) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_training.py:297]: not sure I understand. 
the results of this analysis are outputted to the log alongside with the cross_validation results. 
altogether they suppose to suggest if there is actual overfitting.

[the remark this answers is not in the document]
Comment (TammushB) [reply src/featuremap/ugbio_featuremap/featuremap_xgb_training.py:303]: Ahh got it! adding it.

**Pull request merged**